File: 1.Fundamental-Level/3.OOP/01.Defining-Classes-Homework/01.Persons/Person.cs
using System;

public class Person
{
    public string name;
    public int age;
    public string email;

    public Person(string name, int age)
        : this(name, age, null)
    {
    }

    public Person(string name, int age, string email)
    {
        this.Name = name;
        this.Age = age;
        this.Email = email;
    }

    public string Name
    {
        get
        {
            return this.name;
        }

        set
        {
            if (value == string.Empty)
            {
                throw new ArgumentException("The name cannot be empty string!", "name");
            }

            this.name = value;
        }
    }

    public int Age
    {
        get { return this.age; }
        set
        {
            if (value < 1 || value > 100)
            {
                throw new ArgumentException("The age should be in the range [1... 100]!", "age");
            }

            this.age = value;
        }
    }

    public string Email
    {
        get { return this.email; }
        set
        {
            if (!String.IsNullOrEmpty(value) && !value.Contains("@"))
            {
                throw new ArgumentException("Invalid email!", "email");
            }

            this.email = value;
        }
    }

    public override string ToString()
    {
        string result;

        if (this.Email != null)
        {
            result = string.Format(
            "1.Name -> {0} \n2.Age -> {1} \n3.Email -> {2}",
            this.Name, this.Age, this.Email);
        }
        else
        {
            result = string.Format(
            "1.Name -> {0} \n2.Age -> {1}",
            this.Name, this.Age);
        }

        return result;
    }
}

File: Java-OOP-Basics/Java-OOP-Basics-Solutions/src/encapsulation/_04_ShoppingSpree/Person.java
package encapsulation._04_ShoppingSpree;

import java.util.ArrayList;
import java.util.List;

public class Person {
    private String name;
    private Double money;
    private List<Product> bagOfProducts;

    public Person(String name, Double money) {
        this.setName(name);
        this.setMoney(money);
        this.bagOfProducts = new ArrayList<>();
    }

    private void setName(String name) {
        if (name == null || name.isEmpty() || name.equals(" ")) {
            throw new IllegalArgumentException("Name cannot be empty");
        }

        this.name = name;
    }

    private void setMoney(Double money) {
        if (money < 0) {
            throw new IllegalArgumentException("Money cannot be negative");
        }

        this.money = money;
    }

    public String buy(Product product) {
        String result = "";
        Double productCost = product.getCost();
        if (this.money >= productCost) {
            this.money -= productCost;
            this.bagOfProducts.add(product);
            result = String.format("%s bought %s", this.name, product.getName());
        } else {
            result = String.format("%s can't afford %s", this.name, product.getName());
        }

        return result;
    }

    @Override
    public String toString() {
        StringBuilder helpBuilder = new StringBuilder();
        helpBuilder.append(this.name);
        helpBuilder.append(" - ");
        if (this.bagOfProducts.size() >= 1) {
            for (int i = 0; i < this.bagOfProducts.size() - 1; i++) {
                helpBuilder.append(this.bagOfProducts.get(i).getName()).append(", ");
            }
            helpBuilder.append(this.bagOfProducts.get(this.bagOfProducts.size() - 1).getName());
        } else {
            helpBuilder.append("Nothing bought");
        }

        return helpBuilder.toString();
    }
}